### PR TITLE
Remove auto-cache mechanism

### LINKED ISSUE #1604: Clean up our auto-caching
Currently, some of our trainers cache the data prior to training, without a possibility to disable that. 

I believe the good incremental step would be to disable all auto-caching, and rely on user to call `AppendCacheCheckpoint` prior to multi-pass training. 

This is not really ideal, since the default setup for multi-pass trainers will train slower. I still think it is better to have a consistent story about our 'smarts' (that is, we have no auto-normalization, no auto-caching and no auto-calibration), and use extensive documentation (and tooling, in the future) to cover these pitfalls.

cc @GalOshri @TomFinley @eerhardt

### PR DESCRIPTION
**Pull request by wschin**
Fixes #1604.

### CODE REVIEW

File: src/Microsoft.ML.Data/Training/TrainerEstimatorBase.cs [hunk 1]
```diff
@@ -132,21 +132,16 @@ protected virtual void CheckLabelCompatible(SchemaShape.Column labelCol)
         protected TTransformer TrainTransformer(IDataView trainSet,
             IDataView validationSet = null, IPredictor initPredictor = null)
         {
-            var cachedTrain = Info.WantCaching ? new CacheDataView(Host, trainSet, prefetch: null) : trainSet;
```
Comment (justinormont) [inline src/Microsoft.ML.Data/Training/TrainerEstimatorBase.cs]: Does this remove auto-caching from just the estimators API, or from the application level, like maml.exe?

Comment (wschin) [reply src/Microsoft.ML.Data/Training/TrainerEstimatorBase.cs]: Mainly for estimators but I expect that in the future, users need to call a specific transform to cache an `IDataView` (mentioned in the original issue).

Comment (justinormont) [inline src/Microsoft.ML.Data/Training/TrainerEstimatorBase.cs]: As requested by @GalOshri in the issue, can we [add documentation](https://github.com/dotnet/machinelearning/issues/1604#issuecomment-440824201)? 

Currently the user will have no method of knowing if a specific learner already does it own form of caching, or won't benefit from caching.

Inline w/ @GalOshri's request, I think this documentation should be required before making this change.

Comment (TomFinley) [reply src/Microsoft.ML.Data/Training/TrainerEstimatorBase.cs]: Let's change the appropriate cookbook samples to illustrate the new pattern with this little caching checkpoint thing.

---
In reply to: [237951473](https://github.com/dotnet/machinelearning/pull/1780#discussion_r237951473) [](ancestors = 237951473)

Comment (justinormont) [reply src/Microsoft.ML.Data/Training/TrainerEstimatorBase.cs]: Yes, updating the example code is a good first step. And we should create a direct list of the components which benefit from caching. This is along with _when_ they benefit, for instance, "a LinearSVM when the number of iterations are greater than 1".

Another route is perhaps a VS checker which look at Info.WantCaching and recommends from there?

File: test/Microsoft.ML.Tests/FeatureContributionTests.cs [hunk 2]
```diff
@@ -71,9 +71,10 @@ public void TestFeatureImportance()
             var srcDV = bldr.GetDataView();
 
             var pipeline = ML.Transforms.Concatenate("Features", "X1", "X2Important", "X3", "X4Rand")
+                .AppendCacheCheckpoint(ML)
                 .Append(ML.Transforms.Normalize("Features"));
             var data = pipeline.Fit(srcDV).Transform(srcDV);
-            var model = ML.Regression.Trainers.OnlineGradientDescent().Fit(data);
```
Comment (wschin) [inline test/Microsoft.ML.Tests/FeatureContributionTests.cs]: Because caching change the behavior of `batch size,` we replace `OnlineGradientDescent` with another linear regressor. This should not affect the goal of this test.

File: test/Microsoft.ML.Tests/FeatureContributionTests.cs [hunk 3]
```diff
@@ -85,19 +86,20 @@ public void TestFeatureImportance()
             var enumerator = output.AsEnumerable<ScoreAndContribution>(Env, true).GetEnumerator();
             ScoreAndContribution row = null;
             var expectedValues = new List<float[]>();
-            expectedValues.Add(new float[4] { 0.15640761F, 1, 0.155862764F, 0.07276783F });
-            expectedValues.Add(new float[4] { 0.09507586F, 1, 0.1835608F, 0.0437548943F });
-            expectedValues.Add(new float[4] { 0.297142357F, 1, 0.2855884F, 0.193529665F });
-            expectedValues.Add(new float[4] { 0.45465675F, 0.8805887F, 0.4031663F, 1 });
-            expectedValues.Add(new float[4] { 0.0595234372F, 0.99999994F, 0.349647522F, 0.137912869F });
+            expectedValues.Add(new float[4] { 0.06319684F, 1, 0.1386623F, 4.46209469E-06F });
```
Comment (wschin) [inline test/Microsoft.ML.Tests/FeatureContributionTests.cs]: This new baseline was produced using master branch.

File: test/Microsoft.ML.Tests/Scenarios/OvaTest.cs [hunk 4]
```diff
@@ -133,10 +133,10 @@ public void OvaLinearSvm()
             });
 
             // Data
-            var data = reader.Read(GetDataPath(dataPath));
+            var data = mlContext.Data.Cache(reader.Read(GetDataPath(dataPath)));
 
             // Pipeline
-            var pipeline = new Ova(mlContext, new LinearSvm(mlContext),  useProbabilities: false);
+            var pipeline = new Ova(mlContext, new LinearSvm(mlContext, numIterations: 100),  useProbabilities: false);
```
Comment (wschin) [inline test/Microsoft.ML.Tests/Scenarios/OvaTest.cs]: Since the shuffled sequence changed, it now takes more iterations to converge to previous solution.

File: docs/samples/Microsoft.ML.Samples/Dynamic/SDCA.cs [hunk 5]
```diff
@@ -38,11 +38,18 @@ public static void SDCA_BinaryClassification()
             // Read the data
             var data = reader.Read(dataFile);
 
+            // ML.NET doesn't cache data set by default. Therefore, if one reads a data set from a file and accesses it many times, it can be slow due to
+            // expensive disk operations. When the considered data can fit into memory, a solution is to cache the data in memory. Caching is especially
```
Comment (justinormont) [inline docs/samples/Microsoft.ML.Samples/Dynamic/SDCA.cs]: Generally for a text sentiment dataset, I expect the runtime will be dominated by the feature engineering in the TextTransform (mostly murmur3 in the dictionary lookup, though code profiling is the final answer). I would think this would remain true with the multiple passes without caching.

```suggestion
            // expensive featurization and disk operations. When the considered data can fit into memory, a solution is to cache the data in memory. Caching is especially
```

File: src/Microsoft.ML.Data/Training/TrainerEstimatorBase.cs (hunk 1 above)
Comment (wschin) [reply src/Microsoft.ML.Data/Training/TrainerEstimatorBase.cs]: A sample and some tests are modified to use those caching functions. Every caching function has at least one test now.

File: src/Microsoft.ML.Data/Training/TrainerEstimatorBase.cs [hunk 6]
```diff
@@ -132,21 +132,16 @@ protected virtual void CheckLabelCompatible(SchemaShape.Column labelCol)
...
 
             if (validationSet == null)
-                validRoles = null;
+                validRoleMapped = null;
             else
-            {
-                var cachedValid = Info.WantCaching ? new CacheDataView(Host, validationSet, prefetch: null) : validationSet;
-                validRoles = MakeRoles(cachedValid);
-            }
+                validRoleMapped = MakeRoles(validationSet);
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Data/Training/TrainerEstimatorBase.cs]: just set it null as default and change only if validation set is != null

File: test/Microsoft.ML.Tests/Scenarios/Api/CookbookSamples/CookbookSamples.cs [hunk 7]
```diff
@@ -562,6 +562,9 @@ private void CrossValidationOn(string dataPath)
                     Label: r.Label.ToKey(),
                     // Concatenate all the features together into one column 'Features'.
                     Features: r.SepalLength.ConcatWith(r.SepalWidth, r.PetalLength, r.PetalWidth)))
+                // Add a step for caching data in memory so that the downstream iterative training
+                // algorithm can efficiently scan through the data multiple times.
+                .AppendCacheCheckpoint()
```
Comment (Ivanidzo4ka) [inline test/Microsoft.ML.Tests/Scenarios/Api/CookbookSamples/CookbookSamples.cs]: >AppendCacheCheckpoint [](start = 17, length = 21)

you need to update [https://github.com/dotnet/machinelearning/blob/master/docs/code/MlNetCookBook.md](https://github.com/dotnet/machinelearning/blob/master/docs/code/MlNetCookBook.md) with these changes. (here and in SamplesDynamicApi)

File: test/Microsoft.ML.Tests/TrainerEstimators/SdcaTests.cs [hunk 8]
```diff
@@ -16,7 +16,7 @@ public void SdcaWorkout()
             var dataPath = GetDataPath("breast-cancer.txt");
 
             var data = TextLoader.CreateReader(Env, ctx => (Label: ctx.LoadFloat(0), Features: ctx.LoadFloat(1, 10)))
-                .Read(dataPath);
+                .Read(dataPath).Cache();
```
[the remark this answers is not in the document]
Comment (Ivanidzo4ka) [reply test/Microsoft.ML.Tests/TrainerEstimators/SdcaTests.cs]: What would be user experience if you don't shuffle? Exception?

---
In reply to: [238761329](https://github.com/dotnet/machinelearning/pull/1780#discussion_r238761329) [](ancestors = 238761329)

[the remark this answers is not in the document]
Comment (wschin) [reply test/Microsoft.ML.Tests/TrainerEstimators/SdcaTests.cs]: It can be 100x slower in the worst case.

---
In reply to: [238883768](https://github.com/dotnet/machinelearning/pull/1780#discussion_r238883768) [](ancestors = 238883768,238761329)

File: src/Microsoft.ML.Data/Training/TrainerEstimatorBase.cs (hunk 1 above)
Comment (wschin) [reply src/Microsoft.ML.Data/Training/TrainerEstimatorBase.cs]: I don't think having a list is a small task. We need another PR.

---
In reply to: [237953879](https://github.com/dotnet/machinelearning/pull/1780#discussion_r237953879) [](ancestors = 237953879)

Comment (wschin) [reply src/Microsoft.ML.Data/Training/TrainerEstimatorBase.cs]: Ok. I will do it in next iteration.

---
In reply to: [237951764](https://github.com/dotnet/machinelearning/pull/1780#discussion_r237951764) [](ancestors = 237951764,237951473)

Comment (justinormont) [reply src/Microsoft.ML.Data/Training/TrainerEstimatorBase.cs]: Generally, I like to see documentation in the PR. This is more so true when the user can be surprised by the change and not understand what's different.

Comment (wschin) [reply src/Microsoft.ML.Data/Training/TrainerEstimatorBase.cs]: Many uses are added into Coookbook.

---
In reply to: [238949665](https://github.com/dotnet/machinelearning/pull/1780#discussion_r238949665) [](ancestors = 238949665)

File: test/Microsoft.ML.Tests/Scenarios/Api/CookbookSamples/CookbookSamples.cs (hunk 7 above)
Comment (wschin) [reply test/Microsoft.ML.Tests/Scenarios/Api/CookbookSamples/CookbookSamples.cs]: Done. Could you take a look again?

---
In reply to: [238883605](https://github.com/dotnet/machinelearning/pull/1780#discussion_r238883605) [](ancestors = 238883605)
